targets/ublk.nfs.cpp

```diff
@@ -0,0 +1,413 @@
+/* -*-  mode:c; tab-width:8; c-basic-offset:8; indent-tabs-mode:nil;  -*- */
+// SPDX-License-Identifier: MIT or GPL-2.0-only
+
+#include <config.h>
+
+#include "ublksrv_tgt.h"
+#include <nfsc/libnfs.h>
+
+
+struct nfs_tgt_data {
+        char url[4096];
+	struct nfs_context *nfs;
+        struct nfsfh *nfsfh;
+        size_t capacity;
+};
+
+typedef struct nfs_cb_data {
+        struct nfs_cb_data *next;
+        const struct ublksrv_queue *q;
+        int tag;
+        ssize_t count;
+} nfs_cb_data_t;
+
+struct nfs_queue_data {
+        nfs_cb_data_t *io_list;
+        pthread_spinlock_t io_list_lock;
+        int has_sent_event;
+};
+
+static inline struct nfs_queue_data *
+nfs_get_queue_data(const struct ublksrv_queue *q)
+{
+	return (struct nfs_queue_data *)q->private_data;
+}
+
+
+static void nfs_handle_event(const struct ublksrv_queue *q)
+{
+	struct nfs_queue_data *q_data = nfs_get_queue_data(q);
+        nfs_cb_data_t *cb_data, *tmp;
+        
+	pthread_spin_lock(&q_data->io_list_lock);
+        cb_data = q_data->io_list;
+        q_data->io_list = NULL;
+        q_data->has_sent_event = 0;
+	pthread_spin_unlock(&q_data->io_list_lock);
+
+        while (cb_data) {
+                tmp = cb_data->next;
+
+                ublksrv_complete_io(cb_data->q, cb_data->tag, cb_data->count);
+                free(cb_data);
+                cb_data = tmp;
+        }
+        ublksrv_queue_handled_event(q);
+}
+
+void rw_async_cb(int status, struct nfs_context *nfs,
+                 void *data, void *private_data)
+{
+        nfs_cb_data_t *cb_data = (nfs_cb_data_t *)private_data;
+        const struct ublksrv_queue *q = cb_data->q;
+	struct nfs_queue_data *q_data = nfs_get_queue_data(q);
+
+        if (status < 0) {
+                fprintf(stderr, "pread/pwrite failed with \"%s\"\n", (char *)data);
+                status = -EIO;
+        }
+        cb_data->count = status;
+        
+	pthread_spin_lock(&q_data->io_list_lock);
+        cb_data->next = q_data->io_list;
+        q_data->io_list = cb_data;
+        if (!q_data->has_sent_event) {
+                ublksrv_queue_send_event(q);
```

Comment:
ublksrv_queue_send_event() can be moved out from the spin_lock

Reply:
I moved that out of the spinlock but I now keep  ublksrv_queue_handled_event() and has_seen_event=0 under a spinlock in nfs_handle_event().

There must be a race with managing the has_seen_event variable  but I can not see where the race is.
ublksrv_queue_handled_event() does trigger the event loop to run immediately so that could cause a race? But I can not yet see the race.

My assumption is that there is a race that results in has_seen_event set to 1 after the final nfs_handle_event has completed and thus stopping any future invocations of nfs_handle_event().
And somehow having a spinlock around ublksrv_queue_handle_event()/has_seen_event=0  prevents this race.